android/java/org/chromium/chrome/browser/app/domain/PortfolioModel.java

```diff
@@ -18,23 +18,27 @@
 import org.chromium.brave_wallet.mojom.BlockchainRegistry;
 import org.chromium.brave_wallet.mojom.BlockchainToken;
 import org.chromium.brave_wallet.mojom.BraveWalletService;
+import org.chromium.brave_wallet.mojom.BraveWalletServiceObserver;
 import org.chromium.brave_wallet.mojom.EthTxManagerProxy;
 import org.chromium.brave_wallet.mojom.JsonRpcService;
 import org.chromium.brave_wallet.mojom.KeyringService;
 import org.chromium.brave_wallet.mojom.NetworkInfo;
+import org.chromium.brave_wallet.mojom.OriginInfo;
 import org.chromium.brave_wallet.mojom.SolanaTxManagerProxy;
 import org.chromium.brave_wallet.mojom.TxService;
+import org.chromium.chrome.browser.crypto_wallet.observers.WeakBraveWalletServiceObserver;
 import org.chromium.chrome.browser.crypto_wallet.util.AssetUtils;
 import org.chromium.chrome.browser.crypto_wallet.util.AsyncUtils;
 import org.chromium.chrome.browser.crypto_wallet.util.JavaUtils;
 import org.chromium.chrome.browser.crypto_wallet.util.PortfolioHelper;
+import org.chromium.mojo.system.MojoException;
 
 import java.io.Serializable;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 
-public class PortfolioModel {
+public class PortfolioModel implements BraveWalletServiceObserver {
```

Comment:
I don't really get why we have now 2 classes that `implements` `BraveWalletServiceObserver`, it is confused at least, as `ProfileModel` doesn't really implement it, it does extend `WeakBraveWalletServiceObserver` as we don't add `PortfolioModel` as an observer, we add it as a delegate to `WeakBraveWalletServiceObserver` that really observes.

Reply:
actually it shouldn't even extend `WeakBraveWalletServiceObserver`, the `WeakBraveWalletServiceObserver` just has to pass it's state through a delegate. That's really confusion to have `implements BraveWalletServiceObserver` on 2 classes. The whole goal is to hide `BraveWalletServiceObserver` under `BraveWalletServiceObserverImpl` or `WeakBraveWalletServiceObserverImpl` and don't expose it anywhere else

Reply:
Plus to all the above, creating `PortfolioModel` object leads to creating another `BraveWalletServiceObserver` which does init some proxy stuff and etc, we just need methods there, nothing related to mojo.

Reply:
we don't want that `public interface BraveWalletServiceObserver extends org.chromium.mojo.bindings.Interface {` inside  `PortfolioModel`

Reply:
Yeah, i didn't look too much into the interface. Will add an internal interface.

Reply:
BTW, implementing the `BraveWalletServiceObserver` directly in model shouldn't cause a memory leak as long as we are not passing the object reference to `WalletService` via `BraveWalletService#addObserver`.
I wanted a simple solution but with the strong opinion of wrapper interfaces, seems like we will only be using the wrappers going forward.

Reply:
Yeah it shouldn't add any memory leaks, but who knows what will be added to the upstream's interface in the future